src/lib.rs

```diff
@@ -513,26 +518,29 @@ impl Verification for VerifyOnly {}
 impl Verification for All {}
 
 /// The secp256k1 engine, used to execute all signature operations
-pub struct Secp256k1<C> {
+pub struct Secp256k1<'buf, C> {
     ctx: *mut ffi::Context,
-    phantom: PhantomData<C>
+    phantom: PhantomData<C>,
+    _buf: Option<&'buf [u8]>,
 }
 
 // The underlying secp context does not contain any references to memory it does not own
-unsafe impl<C> Send for Secp256k1<C> {}
+unsafe impl<'buf, C> Send for Secp256k1<'buf, C> {}
 // The API does not permit any mutation of `Secp256k1` objects except through `&mut` references
-unsafe impl<C> Sync for Secp256k1<C> {}
+unsafe impl<'buf, C> Sync for Secp256k1<'buf, C> {}
 
-impl<C> Clone for Secp256k1<C> {
-    fn clone(&self) -> Secp256k1<C> {
+#[cfg(feature = "std")]
+impl<'buf, C> Clone for Secp256k1<'buf, C> {
+    fn clone(&self) -> Secp256k1<'static, C> {
         Secp256k1 {
             ctx: unsafe { ffi::secp256k1_context_clone(self.ctx) },
-            phantom: self.phantom
+            phantom: self.phantom,
+            _buf: None,
         }
     }
 }
```

Comment:
...then I'd implement `Clone` only for the non-`Buffer` versions of `Secp256k1`.